File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
input = sys.stdin.readline

n, m, v = map(int, input().split())
graph = [[] for _ in range(n+1)]

for i in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
for i in range(n+1):
    graph[i].sort()


def dfs(start):
    print(start, end=' ')
    visited1[start] = True

    for i in graph[start]:
        if not visited1[i]:
            dfs(i)


def bfs(start):
    queue = deque([start])
    visited2[start] = True

    while queue:
        node = queue.popleft()
        print(node, end=' ')
        for i in graph[node]:
            if not visited2[i]:
                visited2[i] = True
                queue.append(i)


visited1 = [False for _ in range(n+1)]
visited2 = [False for _ in range(n+1)]
dfs(v)
print()
bfs(v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [n, m, v] = input[0].split(" ").map(Number)
const graph = Array.from({ length: n + 1 }, () => [])
const visitedDfs = Array.from({ length: n + 1 }, () => false)
const visitedBfs = Array.from({ length: n + 1 }, () => false)

input.slice(1).forEach((row) => {
  const [x, y] = row.split(" ").map(Number)
  graph[x].push(y)
  graph[y].push(x)
})

graph.forEach((row) => row.sort((a, b) => a - b))

const dfsAnswer = [v]
const bfsAnswer = [v]

const dfs = (start) => {
  visitedDfs[start] = true

  for (let next of graph[start]) {
    if (!visitedDfs[next]) {
      visitedDfs[next] = true
      dfsAnswer.push(next)
      dfs(next)
    }
  }
}

const bfs = (start) => {
  visitedBfs[start] = true
  const queue = [start]

  while (queue.length > 0) {
    const curr = queue.shift()

    for (let next of graph[curr]) {
      if (!visitedBfs[next]) {
        visitedBfs[next] = true
        bfsAnswer.push(next)
        queue.push(next)
      }
    }
  }
}

dfs(v)
bfs(v)

console.log(dfsAnswer.join(" "))
console.log(bfsAnswer.join(" "))
